File: src/main/ArrayQueue.go
package main

import (
	"fmt"
	"strconv"
)

// go  队列的实现[数组学习  基于切片实现的队列]
func main() {
	queue := initQueue(10)
	for a := 0; a < 10 ; a++ {
		enqueue(queue,strconv.Itoa(a))
	}
	for a := 0; a < 12 ; a++ {
		fmt.Println(dequeue(queue))
	}

}

type QueueStruct struct {
	//队列的数组
	items []string
	//队列的大小
	size int
	//队列的头
	head int
	//队列的尾
	tail int
}

//type IQueue interface {
//	initQueue(size int) *QueueStruct
//	enqueue(queueStruct *QueueStruct, item string) bool
//	dequeue(queueStruct *QueueStruct) string
//}

//初始化队列
func initQueue(size int) *QueueStruct {
	return &QueueStruct{*new([]string),size,0,0,}
}

//入队
func enqueue(queueStruct *QueueStruct, item string) bool {
	if len(queueStruct.items) > queueStruct.size {
		return false
	}
	queueStruct.items = append(queueStruct.items, item)
	queueStruct.tail++
	return true
}

//出队
func dequeue(queueStruct *QueueStruct) string  {
	if queueStruct.head == queueStruct.tail {
		return "队列不为，无元素出队"
	}
	str := queueStruct.items[queueStruct.head]
	queueStruct.head++
	return str

}

File: src/datastructure/queue/ArrayQueue.java
package datastructure.queue;

/**
 * @author huang
 * @version 1.0
 * @date 2019/02/02 11:01
 **/

public class ArrayQueue {
    private String[] items;
    private int n = 0;

    private int head = 0;
    private int tail = 0;

    public ArrayQueue(int capacity) {
        items = new String[capacity];
        n = capacity;
    }

    public boolean enqueue(String item) {
        if (tail == n) {
//            return false;
            if (head == 0) {
                return false;
            }
            for (int i = head; i < tail; ++i) {
                items[i - head] = items[i];
            }
            tail -= head;
            head = 0;
        }
        items[tail] = item;
        ++tail;
        return true;
    }

    public String dequeue() {
        if (head == tail) {
            return null;
        }
        String ret = items[head];
        ++head;
        return ret;
    }

    public static void main(String[] args) {
        ArrayQueue arrayQueue = new ArrayQueue(10);
        for (int i = 0; i < 9; i++) {
            arrayQueue.enqueue(String.valueOf(i));
        }
        for (int i = 0; i < 9; i++) {
            System.out.println(arrayQueue.dequeue());
        }
    }
}
